full review post-trial 03 and tutorials-late (#34)
epiverse-trace/tutorials-late: episodes/simulating-transmission.Rmd

# run an epidemic model using `epidemic()`
output_plot <- epidemics::model_default(
  population = uk_population,
  transmission_rate = transmission_rate,
  infectiousness_rate = infectiousness_rate,
  recovery_rate = recovery_rate,
  time_end = 600, increment = 1.0
)

output_plot %>%
  filter(compartment %in% c("exposed", "infectious")) %>%
  ggplot() +
  geom_line(
    aes(
      x = time,
      y = value,
      colour = demography_group,
      linetype = compartment
    ),
    linewidth = 1.2
  ) +
  scale_y_continuous(
    labels = scales::comma
  ) +
  scale_colour_brewer(
    palette = "Dark2",
    name = "Age group"
  ) +
  theme_bw() +
  labs(
    x = "Simulation time (days)",
    linetype = "Compartment",
    y = "Individuals"
  )
```

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor

By the end of this tutorial, learners should be able to replicate the above image on their own computers.

::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

## Simulating disease spread

To simulate infectious disease trajectories, we must first select a mathematical model to use.

> There is a library of models to choose from in `epidemics`. Models in `epidemics` are prefixed with `model` and suffixed by the name of infection (e.g. Ebola) or a different identifier (e.g. default), and whether the model has a R or [C++](../learners/reference.md#cplusplus) code base.  

Reviewer: Yes, all models are `mode_<NAME>`. There's a case to be made for renaming if that makes a broader use case clear, e.g. `model_diphtheria()` to `model_camps()`.

> There is a library of models to choose from in `epidemics`. Models in `epidemics` are prefixed with `model` and suffixed by the name of infection (e.g. Ebola) or a different identifier (e.g. default), and whether the model has a R or [C++](../learners/reference.md#cplusplus) code base.  

Author: Then these modification seems appropriate:
```suggestion
There is a library of models to choose from in `{epidemics}`. These are prefixed with `model_*` and suffixed by the name of infection (e.g. `model_ebola` for Ebola) or a different identifier (e.g. `model_default`).
```

```{r diagram, echo = FALSE, message = FALSE}
DiagrammeR::grViz("digraph {

  # graph statement
  #################
  graph [layout = dot,
         rankdir = LR,
         overlap = true,
         fontsize = 10]

  # nodes
  #######
  node [shape = square,
       fixedsize = true
       width = 1.3]

       S
       E
       I
       R

  # edges
  #######
  S -> E [label = ' infection \n(transmission rate &beta;)']
  E -> I [label = ' onset of infectiousness \n(infectiousness rate &alpha;)']
  I -> R [label = ' recovery \n(recovery rate &gamma;)']

}")
```

::::::::::::::::::::::::::::::::::::: callout
### Model parameters: rates

In ODE models, model parameters are often (but not always) specified as rates. The rate at which an event occurs is the inverse of the average time until that event. For example, in the SEIR model, the recovery rate $\gamma$ is the inverse of the average infectious period. 

Values of these rates can be determined from the natural history of the disease. For example,  if the average infectious period of an infection is 8 days, then the daily recovery rate is $\gamma = 1/8 = 0.125$.

::::::::::::::::::::::::::::::::::::::::::::::::

> For each disease state ($S$, $E$, $I$ and $R$) and age group ($i$), we have an ordinary differential equation describing the rate of change with respect to time.  

Author: Agree to simplify this.
```suggestion
For each disease state ($S$, $E$, $I$ and $R$) and age group ($i$), we have a differential equation describing the rate of change with respect to time.  
```

Our model output consists of the number of individuals in each compartment in each age group through time. We can visualise the infectious individuals only (those in the $I$ class) through time.

```{r visualise, fig.width = 10}
library(tidyverse)

output %>%
  filter(compartment == "infectious") %>%
  ggplot() +
  geom_line(
    aes(
      x = time,
      y = value,
      colour = demography_group
    )
  ) +
  scale_y_continuous(
    labels = scales::comma
  ) +
  theme_bw() +
  labs(
    x = "Simulation time (days)",
    linetype = "Compartment",
    y = "Individuals"
  )
```

::::::::::::::::::::::::::::::::::::: callout
### Time increments

Note that there is a default argument of `increment = 1`. This relates to the time step of the ODE solver. When the parameters are specified on a daily time-scale and maximum number of time steps (`time_end`) is days, the default time step of the ODE solver one day. 

> The choice of increment will depend on the time scale of the parameters, and the rate at which events can occur. In general, the increment should smaller than the fastest event. For example, if parameters are on a monthly time scale, but some events will occur within a month, then the increment should be less than one month.

Reviewer: It's actually both in {epidemics} as I'm using a simple RK4 stepper and the [integrator uses the `increment` in the observer (output timescale) as well as the stepper](https://www.boost.org/doc/libs/1_85_0/boost/numeric/odeint/integrate/integrate_const.hpp) if the stepper is not an adaptive one. This has thrown up issues in this trial around vaccination and is being tracked on the {epidemics} repo as https://github.com/epiverse-trace/epidemics/issues/226; the fix will be to pick a different one.

The potential blocker there is that the more complex steppers have specific initial state type requirements that involve additional compiled code dependencies, so hopefully I'll find a way to get them to work with current dependencies.

> The choice of increment will depend on the time scale of the parameters, and the rate at which events can occur. In general, the increment should smaller than the fastest event. For example, if parameters are on a monthly time scale, but some events will occur within a month, then the increment should be less than one month.

Author: > Could also rephrase the example time scale to be consistent with flu, e.g. cases are reported daily, makes sense to output daily event data?

We can compare two scenarios in examples:

```suggestion
The choice of increment will depend on the time scale of the parameters, and the rate at which events can occur. In general, the increment should be smaller than the fastest event. For example: 

- if parameters are on a daily time scale, and all events will occur within a day, then the increment should be equal to one day;
- if parameters are on a monthly time scale, but some events will occur within a month, then the increment should be less than one month.
```
